lib/streamlit/version.py

```diff
@@ -0,0 +1,100 @@
+# Copyright 2018-2019 Streamlit Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Streamlit version utilities."""
+import random
+
+import pkg_resources
+import requests
+
+from streamlit.logger import get_logger
+
+LOGGER = get_logger(__name__)
+
+PYPI_STREAMLIT_URL = 'https://pypi.org/pypi/streamlit/json'
+
+# Probability that we'll make a network call to PyPI to check
+# the latest version of streamlit. This is used each time
+# should_show_new_version_notice() is called.
+CHECK_PYPI_PROBABILITY = 0.05
+
+
+def get_installed_streamlit_version():
+    """Return the streamlit version string from setup.py.
+
+    Returns
+    -------
+    str
+        The version string specified in setup.py.
+
+    """
+    return pkg_resources.get_distribution('streamlit').version
+
+
+def get_latest_streamlit_version(timeout=None):
+    """Request the latest streamlit version string from PyPI.
+
+    NB: this involves a network call, so it could raise an error
+    or take a long time.
+
+    Parameters
+    ----------
+    timeout : float | None
+        The request timeout.
+
+    Returns
+    -------
+    str
+        The version string for the latest version of streamlit
+        on PyPI.
+
+    """
+    rsp = requests.get(PYPI_STREAMLIT_URL, timeout=timeout)
+    try:
+        return rsp.json()['info']['version']
+    except Exception as e:
+        raise RuntimeError('Got unexpected response from PyPI', e)
+
+
+def should_show_new_version_notice():
+    """True if streamlit should show a 'new version!' notice to the user.
+
+    We need to make a network call to PyPI to determine the latest streamlit
+    version. Since we don't want to do this every time streamlit is run,
+    we'll only perform the check ~5% of the time.
+
+    If we do make the request to PyPI and there's any sort of error,
+    we log it and return False.
+
+    Returns
+    -------
+    bool
+        True if we should tell the user that their streamlit is out of date.
+
+    """
+    if random.random() >= CHECK_PYPI_PROBABILITY:
+        # We don't check PyPI every time this function is called.
+        LOGGER.debug('Skipping PyPI version check')
+        return False
+
+    try:
+        installed_version = get_installed_streamlit_version()
+        latest_version = get_latest_streamlit_version(timeout=1)
+    except BaseException as e:
+        # Log this as a debug. We don't care if the user sees it.
+        LOGGER.debug('Failed PyPI version check', e)
```

Comment:
This notation doesn't actually work. I think you meant something like this:

```
LOGGER.debug('Failed PyPI version check.\n%s', e)
```

Otherwise you get:
```
Traceback (most recent call last):
  File "/usr/lib/python2.7/logging/__init__.py", line 868, in emit
    msg = self.format(record)
  File "/usr/lib/python2.7/logging/__init__.py", line 741, in format
    return fmt.format(record)
  File "/usr/lib/python2.7/logging/__init__.py", line 465, in format
    record.message = record.getMessage()
  File "/usr/lib/python2.7/logging/__init__.py", line 329, in getMessage
    msg = msg % self.args
TypeError: not all arguments converted during string formatting
```